File: app/src/main/java/com/voiceassist/lixinyu/voiceassist/settings/ui/EditNodeActivity.java
package com.voiceassist.lixinyu.voiceassist.settings.ui;

import android.animation.Animator;
import android.animation.AnimatorInflater;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.view.ViewStub;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.Button;
import android.widget.ImageView;

import com.voiceassist.lixinyu.voiceassist.R;
import com.voiceassist.lixinyu.voiceassist.common.BaseActivity;
import com.voiceassist.lixinyu.voiceassist.common.Constants;
import com.voiceassist.lixinyu.voiceassist.common.widget.RecyclerViewDivider;
import com.voiceassist.lixinyu.voiceassist.common.widget.dialog.CommonContentDialog;
import com.voiceassist.lixinyu.voiceassist.entity.dto.Node;
import com.voiceassist.lixinyu.voiceassist.entity.dto.Relationship;
import com.voiceassist.lixinyu.voiceassist.entity.dto.SecondLevelNode;
import com.voiceassist.lixinyu.voiceassist.main.ui.MainActivity;
import com.voiceassist.lixinyu.voiceassist.settings.adapter.NodeListAdapter;
import com.voiceassist.lixinyu.voiceassist.utils.ToastUtils;

import java.io.File;
import java.util.List;

/**
 * Created by lilidan on 2018/1/25.
 */

public class EditNodeActivity extends BaseActivity implements IEmptyable {

    private static final int REQ_ADD_EDIT = 0x100;

    private ImageView mBtnAddNode;

    private RecyclerView mRecyclerView;
    private NodeListAdapter mAdapter;

    private List<Node> mNodeList;

    private CommonContentDialog mDeleteDialog;

    private Animation mAnimOut;
    private Animation mAnimIn;
    private boolean mAlreayOut;

    private ViewStub mEmptyViewStub;


    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_edit_node);

        initView();
        initData();
        initListener();
    }

    private void initView() {
        mBtnAddNode = findViewById(R.id.edit_node_add_button);
        mRecyclerView = findViewById(R.id.edit_node_recyclerview);
        mEmptyViewStub = findViewById(R.id.edit_node_empty_view_stub);
    }

    private void initData() {
        setTitle("添加或编辑节点");

        mNodeList = MainActivity.Companion.getMAllData().nodes;

        mAdapter = new NodeListAdapter(EditNodeActivity.this, mNodeList);

        mRecyclerView.setLayoutManager(new LinearLayoutManager(EditNodeActivity.this, LinearLayoutManager.VERTICAL, false));
        mRecyclerView.addItemDecoration(new RecyclerViewDivider(EditNodeActivity.this, LinearLayoutManager.HORIZONTAL));
        mRecyclerView.setItemAnimator(new DefaultItemAnimator());
        mRecyclerView.setAdapter(mAdapter);

        justifyDisplayEmptyView();

        mAnimOut = AnimationUtils.loadAnimation(this, R.anim.out_to_bottom);
        mAnimIn = AnimationUtils.loadAnimation(this, R.anim.in_from_bottom);

        ToastUtils.showToast("长按可删除");
    }

    private void initListener() {
        mAdapter.setOnItemClickListener(new NodeListAdapter.OnItemClickListener() {
            @Override
            public void onClick(NodeListAdapter adapter, int position, Node node) {
                Intent intent = new Intent(EditNodeActivity.this, NodeAddEditActivity.class);
                intent.putExtra("position", position);
                intent.putExtra("node", node);
                startActivityForResult(intent, REQ_ADD_EDIT);
            }

            @Override
            public void onLongClick(NodeListAdapter adapter, final int position, final Node node) {
                if (null == mDeleteDialog) {
                    mDeleteDialog = new CommonContentDialog.Builder(EditNodeActivity.this)
                            .contentText("确认删除" + node.cnName + "节点吗？")
                            .onYesClickListener(new View.OnClickListener() {
                                @Override
                                public void onClick(View v) {
                                    Object objPos = v.getTag(R.id.node_delete_position_tag);
                                    Object objNode = v.getTag(R.id.node_delete_entity_tag);

                                    int pos = -1;
                                    if (null != objPos && objPos instanceof Integer) {
                                        pos = (int) objPos;
                                    } else {
                                        pos = position;
                                    }

                                    Node realNode = null;
                                    if (null != objNode && objNode instanceof Node) {
                                        realNode = (Node) objNode;
                                    } else {
                                        realNode = node;
                                    }

                                    mDeleteDialog.dismiss();

                                    // start delete
                                    mNodeList.remove(pos);
                                    mAdapter.notifyDataSetChanged();
                                    justifyDisplayEmptyView();

                                    if (null != realNode) {
                                        deleteRelationship(realNode);
                                        deleteVoiceFile(realNode.audioPath);
                                    }

                                    MainActivity.Companion.saveAllDatas();
                                }
                            })
                            .onNoClickListener(new View.OnClickListener() {
                                @Override
                                public void onClick(View v) {
                                    mDeleteDialog.dismiss();
                                }
                            })
                            .build();
                } else {
                    mDeleteDialog.getContentTextView().setText("确认删除“" + node.cnName + "”节点吗？");
                }

                Button yesButton = mDeleteDialog.getYesButton();
                if (null != yesButton) {
                    yesButton.setTag(R.id.node_delete_position_tag, position);
                    yesButton.setTag(R.id.node_delete_entity_tag, node);
                }
                mDeleteDialog.show();
            }
        });

        mBtnAddNode.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivityForResult(new Intent(EditNodeActivity.this, NodeAddEditActivity.class), REQ_ADD_EDIT);
            }
        });


        mRecyclerView.setOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrollStateChanged(RecyclerView recyclerView, int newState) {
                super.onScrollStateChanged(recyclerView, newState);
            }

            @Override
            public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
                if (dy <= 0) {
                    if (mAlreayOut) {
                        mBtnAddNode.startAnimation(mAnimIn);
                        mAlreayOut = false;
                        mBtnAddNode.setClickable(true);
                    }
                } else {
                    if (!mAlreayOut) {
                        mBtnAddNode.startAnimation(mAnimOut);
                        mAlreayOut = true;
                        mBtnAddNode.setClickable(false);
                    }
                }
            }
        });


    }

    private void deleteRelationship(Node node) {
        if (null == node) return;

        String deleteId = node.id;
        MainActivity.Companion.getMNodesMap().remove(deleteId);

        // 删除关系列表中相关内容
        List<Relationship> relationshipList = MainActivity.Companion.getMAllData().relationship;
        if (null != relationshipList) {
            for (int i = 0, level1Size = relationshipList.size(); i < level1Size;) {
                Relationship relationship = relationshipList.get(i);
                if (null == relationship) {
                    i++;
                    continue;
                }
                if (relationship.firstLevelNodeId.equals(deleteId)) {
                    relationshipList.remove(i);
                    level1Size--;
                    continue;
                } else {
                    List<SecondLevelNode> secondLevelNodeList = relationship.secondLevelNodes;
                    if (null != secondLevelNodeList) {
                        for (int j = 0, level2Size = secondLevelNodeList.size(); j < level2Size;) {
                            SecondLevelNode secondLevelNode = secondLevelNodeList.get(j);
                            if (null == secondLevelNode) {
                                j++;
                                continue;
                            }
                            if (secondLevelNode.secondLevelNodeId.equals(deleteId)) {
                                secondLevelNodeList.remove(j);
                                level2Size--;
                                continue;
                            }
                            j++;
                        }
                    }
                }
                i++;
            }
        }
    }

    private void deleteVoiceFile(String filePath) {
        if (null == filePath || filePath.length() == 0) return;

        if (!filePath.contains(Constants.ROOT_PATH)) {
            filePath = Constants.ROOT_PATH + filePath;
        }

        File file = new File(filePath);
        if (file.exists()) {
            file.delete();
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (RESULT_OK != resultCode) return;

        switch (requestCode) {
            case REQ_ADD_EDIT: {
                int position = data.getIntExtra("position", -1);
                Node node = (Node) data.getSerializableExtra("node");

                if (position != -1) {
                    mNodeList.set(position, node);
                } else {
                    mNodeList.add(0, node);
                }

                MainActivity.Companion.getMNodesMap().put(node.id, node);
                mAdapter.notifyDataSetChanged();
                justifyDisplayEmptyView();
                MainActivity.Companion.saveAllDatas();
                break;
            }

            default: {
                break;
            }
        }
    }

    @Override
    public void justifyDisplayEmptyView() {
        if (null == mNodeList || mNodeList.size() == 0) {
            mEmptyViewStub.setVisibility(View.VISIBLE);
        } else {
            mEmptyViewStub.setVisibility(View.GONE);
        }
    }
}

File: app/src/main/java/com/voiceassist/lixinyu/voiceassist/settings/ui/EditNodeActivity.kt
package com.voiceassist.lixinyu.voiceassist.settings.ui

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.support.v7.widget.DefaultItemAnimator
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.View
import android.view.ViewStub
import android.view.animation.Animation
import android.view.animation.AnimationUtils
import android.widget.ImageView
import com.voiceassist.lixinyu.voiceassist.R
import com.voiceassist.lixinyu.voiceassist.common.BaseActivity
import com.voiceassist.lixinyu.voiceassist.common.Constants
import com.voiceassist.lixinyu.voiceassist.common.widget.RecyclerViewDivider
import com.voiceassist.lixinyu.voiceassist.common.widget.dialog.CommonContentDialog
import com.voiceassist.lixinyu.voiceassist.entity.dto.Node
import com.voiceassist.lixinyu.voiceassist.main.ui.MainActivity
import com.voiceassist.lixinyu.voiceassist.settings.adapter.NodeListAdapter
import com.voiceassist.lixinyu.voiceassist.utils.ToastUtils
import java.io.File

/**
 * Created by lilidan on 2018/1/25.
 */

class EditNodeActivity : BaseActivity(), IEmptyable {

    private var mBtnAddNode: ImageView? = null

    private var mRecyclerView: RecyclerView? = null
    private var mAdapter: NodeListAdapter? = null

    private var mNodeList: MutableList<Node>? = null

    private var mDeleteDialog: CommonContentDialog? = null

    private var mAnimOut: Animation? = null
    private var mAnimIn: Animation? = null
    private var mAlreayOut: Boolean = false

    private var mEmptyViewStub: ViewStub? = null


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_edit_node)

        initView()
        initData()
        initListener()
    }

    private fun initView() {
        mBtnAddNode = findViewById(R.id.edit_node_add_button)
        mRecyclerView = findViewById(R.id.edit_node_recyclerview)
        mEmptyViewStub = findViewById(R.id.edit_node_empty_view_stub)
    }

    private fun initData() {
        title = "添加或编辑节点"

        mNodeList = MainActivity.mAllData!!.nodes

        mAdapter = NodeListAdapter(this@EditNodeActivity, mNodeList)

        mRecyclerView!!.layoutManager = LinearLayoutManager(this@EditNodeActivity, LinearLayoutManager.VERTICAL, false)
        mRecyclerView!!.addItemDecoration(RecyclerViewDivider(this@EditNodeActivity, LinearLayoutManager.HORIZONTAL))
        mRecyclerView!!.itemAnimator = DefaultItemAnimator()
        mRecyclerView!!.adapter = mAdapter

        justifyDisplayEmptyView()

        mAnimOut = AnimationUtils.loadAnimation(this, R.anim.out_to_bottom)
        mAnimIn = AnimationUtils.loadAnimation(this, R.anim.in_from_bottom)

        ToastUtils.showToast("长按可删除")
    }

    private fun initListener() {
        mAdapter!!.setOnItemClickListener(object : NodeListAdapter.OnItemClickListener {


            override var onClick: ((NodeListAdapter, Int, Node?) -> Unit)? = {
                adapter, position, node ->

                val intent = Intent(this@EditNodeActivity, NodeAddEditActivity::class.java)
                intent.putExtra("position", position)
                intent.putExtra("node", node)
                startActivityForResult(intent, REQ_ADD_EDIT)
            }

            override var onLongClick: ((NodeListAdapter, Int, Node?) -> Unit)? = {
                adapter, position, node ->

                if (mDeleteDialog == null) {
                    mDeleteDialog = CommonContentDialog.Builder(this@EditNodeActivity)
                            .contentText("确认删除" + node?.cnName + "节点吗？")
                            .onYesClickListener { v ->
                                val objPos = v.getTag(R.id.node_delete_position_tag)
                                val objNode = v.getTag(R.id.node_delete_entity_tag)

                                var pos = -1
                                if (null != objPos && objPos is Int) {
                                    pos = objPos
                                } else {
                                    pos = position
                                }

                                var realNode: Node? = null
                                if (null != objNode && objNode is Node) {
                                    realNode = objNode
                                } else {
                                    realNode = node
                                }

                                mDeleteDialog!!.dismiss()

                                // start delete
                                mNodeList!!.removeAt(pos)
                                mAdapter!!.notifyDataSetChanged()
                                justifyDisplayEmptyView()

                                if (null != realNode) {
                                    deleteRelationship(realNode)
                                    deleteVoiceFile(realNode.audioPath)
                                }

                                MainActivity.saveAllDatas()
                            }
                            .onNoClickListener { mDeleteDialog!!.dismiss() }
                            .build()
                } else {
                    mDeleteDialog?.contentTextView?.text = "确认删除“" + node?.cnName + "”节点吗？"
                }

                val yesButton = mDeleteDialog!!.yesButton
                if (null != yesButton) {
                    yesButton.setTag(R.id.node_delete_position_tag, position)
                    yesButton.setTag(R.id.node_delete_entity_tag, node)
                }
                mDeleteDialog!!.show()
            }
        })

        mBtnAddNode!!.setOnClickListener { startActivityForResult(Intent(this@EditNodeActivity, NodeAddEditActivity::class.java), REQ_ADD_EDIT) }


        mRecyclerView!!.setOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrollStateChanged(recyclerView: RecyclerView?, newState: Int) {
                super.onScrollStateChanged(recyclerView, newState)
            }

            override fun onScrolled(recyclerView: RecyclerView?, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)
                if (dy <= 0) {
                    if (mAlreayOut) {
                        mBtnAddNode!!.startAnimation(mAnimIn)
                        mAlreayOut = false
                        mBtnAddNode!!.isClickable = true
                    }
                } else {
                    if (!mAlreayOut) {
                        mBtnAddNode!!.startAnimation(mAnimOut)
                        mAlreayOut = true
                        mBtnAddNode!!.isClickable = false
                    }
                }
            }
        })


    }

    private fun deleteRelationship(node: Node?) {
        if (null == node) return

        val deleteId = node.id
        MainActivity.mNodesMap!!.remove(deleteId)

        // 删除关系列表中相关内容
        val relationshipList = MainActivity.mAllData!!.relationship
        if (null != relationshipList) {
            var i = 0
            var level1Size = relationshipList.size
            while (i < level1Size) {
                val relationship = relationshipList[i]
                if (null == relationship) {
                    i++
                    continue
                }
                if (relationship.firstLevelNodeId == deleteId) {
                    relationshipList.removeAt(i)
                    level1Size--
                    continue
                } else {
                    val secondLevelNodeList = relationship.secondLevelNodes
                    if (null != secondLevelNodeList) {
                        var j = 0
                        var level2Size = secondLevelNodeList.size
                        while (j < level2Size) {
                            val secondLevelNode = secondLevelNodeList[j]
                            if (null == secondLevelNode) {
                                j++
                                continue
                            }
                            if (secondLevelNode.secondLevelNodeId == deleteId) {
                                secondLevelNodeList.removeAt(j)
                                level2Size--
                                continue
                            }
                            j++
                        }
                    }
                }
                i++
            }
        }
    }

    private fun deleteVoiceFile(filePath: String?) {
        var filePath = filePath
        if (null == filePath || filePath.length == 0) return

        if (!filePath.contains(Constants.ROOT_PATH)) {
            filePath = Constants.ROOT_PATH + filePath
        }

        val file = File(filePath)
        if (file.exists()) {
            file.delete()
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (data == null) return

        if (Activity.RESULT_OK != resultCode) return

        when (requestCode) {
            REQ_ADD_EDIT -> {
                val position = data.getIntExtra("position", -1)
                val node = data.getSerializableExtra("node") as Node

                if (position != -1) {
                    mNodeList!![position] = node
                } else {
                    mNodeList!!.add(0, node)
                }

                MainActivity.mNodesMap!![node.id] = node
                mAdapter!!.notifyDataSetChanged()
                justifyDisplayEmptyView()
                MainActivity.saveAllDatas()
            }

            else -> {
            }
        }
    }

    override fun justifyDisplayEmptyView() {
        if (null == mNodeList || mNodeList!!.size == 0) {
            mEmptyViewStub!!.visibility = View.VISIBLE
        } else {
            mEmptyViewStub!!.visibility = View.GONE
        }
    }

    companion object {

        private val REQ_ADD_EDIT = 0x100
    }
}
